# Projeto 5 - Modelo de Leslie
João Victor Dell Agli Floriano, 10799783

O Modelo de Leslie, introduzido por Patrick H. Leslie em 1945 em seu artigo, *On the use of matrices in certain population mathematics*, apresenta uma maneira de estuturar e modelar o crescimento populacional por idade. O modelo, de comportamento de tempo discreto, se baseia na ideia de fertilidade por faixas de idade, separando a população com útero fértil em faixas, que vão de **0** a **49** anos de idade.

Considerando uma população com faixas de idade separadas de 5 em 5 anos, as quais tais faixas são referidas como $P_{k, t}$, a população da faixa $k$ no tempo $t$. O crescimento de cada faixa aqui é baseado em dois parâmetros:

* $f_k$ : Fertilidade na faixa $k$, sendo $f_k \geq 0$.
* $s_k$ : Probabilidade de um animal passar do tempo $t$ para o tempo $t+1$, definida entre $[0.0, 1.0]$.

A partir desses dois parâmetros, conseguimos organizar um sistema linear que, para cada faixa de idade, existe uma fertilidade e probabilidade de sobrevivência associadas, que se organizam da seguinte maneira:

$\begin{bmatrix} P_{0, t + 1} \\ P_{1, t + 1} \\ \vdots \\ P_{K, t + 1} \end{bmatrix} = \begin{bmatrix} f_0 & f_1 & \cdots & f_K \\ s_0 & 0 &\cdots & 0 \\ \vdots & \vdots & \ddots & \vdots \\ 0 & 0 & \cdots & 0 \end{bmatrix} \begin{bmatrix} P_{0, t} \\ P_{1, t} \\ \vdots \\ P_{K, t} \end{bmatrix}$

Nesse sistema linear, a matriz do meio, que organiza as fertilidades e probabilidades, é a matriz de Leslie.

$\textbf{P}_{t + 1} = \textbf{M} \textbf{P}_t$

Para resolver esse sistema, testemos um chute:

$\textbf{P}_t = r^t \textbf{V}$

$r^{t+1} \textbf{V} = \textbf{M} r^t \textbf{V}$

$r \textbf{V} = \textbf{M} \textbf{V}$

$\textbf{M} \textbf{V} = r \textbf{V}$

Que indica que esse chute é equivalente à solução de um sistema de autovalores e autovetores, portanto, basta calculá-los para encontrar valores de r, e as distribuições estáveis.

O que Leslie provou foi que, calculando esses autovalores, apenas o maior autovalor indicaria o comportamento assintótico de $\textbf{P(t)}$.

## Implementação

In [2]:
import numpy as np

Para a Matriz de Leslie de exemplo:

In [3]:
leslie = np.array([[0.00000, 0.00105, 0.08203, 0.28849, 0.37780, 0.26478, 0.14055, 0.05857, 0.01344, 0.00081],
                   [0.99694, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                   [0.0, 0.99842, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                   [0.0, 0.0, 0.99785, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                   [0.0, 0.0, 0.0, 0.99671, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                   [0.0, 0.0, 0.0, 0.0, 0.99614, 0.0, 0.0, 0.0, 0.0, 0.0],
                   [0.0, 0.0, 0.0, 0.0, 0.0, 0.99496, 0.0, 0.0, 0.0, 0.0],
                   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.99247, 0.0, 0.0, 0.0],
                   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.98875, 0.0, 0.0],
                   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.98305, 0.0]])

Calculemos os autovalores e autovetores:

In [10]:
val, vec = np.linalg.eig(leslie)

print(val)
print(vec)

[ 1.03765055+0.j          0.30981229+0.73740434j  0.30981229-0.73740434j
  0.01057436+0.51101901j  0.01057436-0.51101901j -0.3879696 +0.37153279j
 -0.3879696 -0.37153279j -0.40819376+0.10637076j -0.40819376-0.10637076j
 -0.08609712+0.j        ]
[[ 3.74609824e-01+0.00000000e+00j -3.55671587e-02-7.59189277e-02j
  -3.55671587e-02+7.59189277e-02j  3.98101929e-04+2.11317904e-03j
   3.98101929e-04-2.11317904e-03j  2.73952324e-03-1.83686145e-03j
   2.73952324e-03+1.83686145e-03j -2.67214306e-04-3.02537556e-04j
  -2.67214306e-04+3.02537556e-04j -2.73592209e-10+0.00000000e+00j]
 [ 3.59912610e-01+0.00000000e+00j -1.04411391e-01+4.21802682e-03j
  -1.04411391e-01-4.21802682e-03j  4.13687176e-03-6.91048604e-04j
   4.13687176e-03+6.91048604e-04j -6.02988393e-03-1.05435819e-03j
  -6.02988393e-03+1.05435819e-03j  4.30820000e-04+8.51160579e-04j
   4.30820000e-04-8.51160579e-04j  3.16799244e-09+0.00000000e+00j]
 [ 3.46305361e-01+0.00000000e+00j -4.56293858e-02+1.22198734e-01j
  -4.56293858e-02-1.2219873

A partir dos autovalores, conseguimos encontrar apenas os valores reais:

In [5]:
val_r = val[val.imag == 0.0]

print(val_r)

[ 1.03765055+0.j -0.08609712+0.j]


E então, procuramos pelo maior valor:

In [6]:
r = val_r[val_r == np.max(val_r)][0].real

print(r)

1.037650550443967


Que, no caso dessa matriz, é $r_1 \approx 1.0376$, que indica que a população de mulheres de idade menor que 50 anos, cresce exponencialmente com o tempo já que $r_1 > 1$, na taxa de $3.76\%$ a cada 5 anos.

A partir desse autovalor, seu autovetor associado que indica a distribuição estável das faixas é:


In [28]:
dist = vec.T[val.imag == 0.0]

dist = dist[val_r == np.max(val_r)].real[0]

dist = dist/np.sum(dist)
print(dist)


[0.11932148 0.11464009 0.11030589 0.10607495 0.10188976 0.09781372
 0.09378951 0.0897058  0.08547831 0.08098049]


Que, traduzido em porcentagens é:

$$D = \begin{bmatrix}11.93\% \\ 11.46 \% \\ 11.03 \% \\ 10.60 \% \\ 10.18 \% \\ 9.78 \% \\ 9.37 \% \\ 8.97 \% \\ 8.54 \% \\ 8.09 \% \end{bmatrix}$$